File: presentations/14-repository-and-builder-pattern-demos/csharp-smart-home-scenes/Builder/SceneDirector.cs
using SmartHomeScenesDemo.Domain;

namespace SmartHomeScenesDemo.Builder;

public sealed class SceneDirector
{
    private readonly DeviceSceneBuilder _builder;

    public SceneDirector(DeviceSceneBuilder builder)
    {
        _builder = builder;
    }

    public DeviceScene BuildEveningArrival(Guid porchLightId)
    {
        _builder.Reset();

        return _builder
            .Named("Evening Arrival")
            .AddDeviceAction(porchLightId, "TurnOn")
            .AddGroupAction("Light", "Living Room", "TurnOn")
            .AddGroupAction(
                "Light",
                "Living Room",
                "SetBrightness",
                new Dictionary<string, string>
                {
                    ["brightness"] = "40"
                })
            .Build();
    }
}

File: presentations/14-repository-and-builder-pattern-demos/java-smart-home-scenes/src/main/java/edu/ksu/swe4743/smarthomescenes/builder/SceneDirector.java
package edu.ksu.swe4743.smarthomescenes.builder;

import edu.ksu.swe4743.smarthomescenes.domain.DeviceScene;
import java.util.Map;
import java.util.UUID;

public final class SceneDirector {
    private final DeviceSceneBuilder builder;

    public SceneDirector(DeviceSceneBuilder builder) {
        this.builder = builder;
    }

    public DeviceScene buildEveningArrival(UUID porchLightId) {
        builder.reset();

        return builder
            .named("Evening Arrival")
            .addDeviceAction(porchLightId, "TurnOn", Map.of())
            .addGroupAction("Light", "Living Room", "TurnOn", Map.of())
            .addGroupAction("Light", "Living Room", "SetBrightness", Map.of("brightness", "40"))
            .build();
    }
}
